Tasks/AzureRmWebAppDeployment/operations/PublishProfileUtility.ts

```diff
@@ -0,0 +1,111 @@
+        msDeployPublishingProfile.publishUrl = this._publishProfileJs.MSDeployServiceURL[0];
+        msDeployPublishingProfile.userName = this._publishProfileJs.UserName[0];
+        msDeployPublishingProfile.userPWD = taskParams.PublishProfilePassword;
+        return msDeployPublishingProfile;
+    }
+
+    public async GetPropertyValuefromPublishProfile(propertyKey: string): Promise<any> {
+        try {
+            if(this._publishProfileJs === null) {
+                this._publishProfileJs = await this.GetPublishProfileJsonFromFile();
+            }
+        } catch(error) {
+            throw new Error(error);
+        }
+        return new Promise ((response, reject) => {
+            this._publishProfileJs.hasOwnProperty(propertyKey) ?
+                response(this._publishProfileJs[propertyKey][0]) : reject(tl.loc('PropertyDoesntExistPublishProfile', propertyKey)); 
+            });
+    }
+
+    private async GetPublishProfileJsonFromFile(): Promise<any> {
+        return new Promise((response, reject) => {
+            var pubxmlFile = packageUtility.PackageUtility.getPackagePath(this._publishProfilePath);
+            var publishProfileXML = fs.readFileSync(pubxmlFile);
+            parseString(publishProfileXML, (error, result) => {
+                if(!!error) {
+                    reject(tl.loc('XmlParsingFailed', error));
+                }
+                var propertyGroup = result && result.Project && result.Project.PropertyGroup ? result.Project.PropertyGroup : null;
+                if(propertyGroup) {
+                    for (var index in propertyGroup) {
+                        if (propertyGroup[index] && propertyGroup[index].WebPublishMethod[0] === Constant.PublishProfileXml.MSDeploy) {
+                            if(!propertyGroup[index].hasOwnProperty(Constant.PublishProfileXml.MSDeployServiceURL) 
+                                || !propertyGroup[index].hasOwnProperty(Constant.PublishProfileXml.DeployIisAppPath) 
+                                || !propertyGroup[index].hasOwnProperty(Constant.PublishProfileXml.UserName)) {
+                                reject(tl.loc('InvalidPublishProfile'));
+                            }
+                            tl.debug("Publish Profile: " + JSON.stringify(propertyGroup[index]));
+                            response(propertyGroup[index]);
+                        }
+                    }
+                }
+                reject(tl.loc('ErrorNoSuchDeployingMethodExists'));
+            });
+        });
+    }
+    
+    public async RunCmd(cmdArgs) {
+        var deferred = Q.defer();
+        var cmdError = null;
+        var errorFile = path.join(tl.getVariable('System.DefaultWorkingDirectory'), ERROR_FILE_NAME);
+		var errObj = fs.createWriteStream(errorFile);
+        errObj.on('finish', () => {
+            if(cmdError) {
+                deferred.reject(cmdError);
+            }
+        });
+
+        try {
```

Comment:
You dont need to append then if you are using await. you can simplify by:
try
{await tl.exec();
deferred.resolve();
}
catch {
cmdError = error;
}

Reply:
done